File: graph/components/bipartite_check.py
import sys
from collections import deque

# Fast I/O
input = sys.stdin.readline
sys.setrecursionlimit(200000)

class BipartiteCheck:
    """
    Check if a graph is Bipartite (2-colorable)
    Time Complexity: O(V + E)
    """
    def __init__(self, n, adj):
        self.n = n
        self.adj = adj
        self.color = [-1] * (n + 1) # -1: uncolored, 0: color A, 1: color B
        self.is_bipartite = True

    def check_bfs(self, start_node):
        """Returns False if not bipartite"""
        q = deque([start_node])
        self.color[start_node] = 0
        
        while q:
            u = q.popleft()
            
            for v in self.adj[u]:
                if self.color[v] == -1:
                    self.color[v] = 1 - self.color[u]
                    q.append(v)
                elif self.color[v] == self.color[u]:
                    self.is_bipartite = False
                    return False
        return True

    def solve(self):
        for i in range(1, self.n + 1):
            if self.color[i] == -1:
                if not self.check_bfs(i):
                    return False
        return True

def solve():
    try:
        line1 = input().split()
        if not line1: return
        n, m = map(int, line1)
        
        adj = [[] for _ in range(n + 1)]
        for _ in range(m):
            u, v = map(int, input().split())
            adj[u].append(v)
            adj[v].append(u)
            
        checker = BipartiteCheck(n, adj)
        if checker.solve():
            print("YES")
            # Print coloring
            # print(checker.color[1:])
        else:
            print("NO")
            
    except ValueError:
        pass

if __name__ == "__main__":
    solve()

File: graph/components/bipartite_check.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Bipartite Check Template
 * Checks if graph can be colored with 2 colors
 * Time Complexity: O(V + E)
 */

const int MAXN = 200005;
vector<int> adj[MAXN];
int color[MAXN]; // -1: uncolored, 0, 1
bool is_bipartite = true;

void bfs_check(int start_node) {
    queue<int> q;
    q.push(start_node);
    color[start_node] = 0;
    
    while (!q.empty()) {
        int u = q.front();
        q.pop();
        
        for (int v : adj[u]) {
            if (color[v] == -1) {
                color[v] = 1 - color[u];
                q.push(v);
            } else if (color[v] == color[u]) {
                is_bipartite = false;
                return;
            }
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n, m;
    if (!(cin >> n >> m)) return 0;
    
    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }
    
    fill(color, color + n + 1, -1);
    
    for (int i = 1; i <= n; i++) {
        if (color[i] == -1) {
            bfs_check(i);
            if (!is_bipartite) break;
        }
    }
    
    if (is_bipartite) cout << "YES\n";
    else cout << "NO\n";
    
    return 0;
}
